### PR TITLE
FS2-2734: fix label weirdness with small multiples

### PR DESCRIPTION
**Pull request by JustinCCYap**
- Trim whitespace that is added to labels for uniqueness
- Fix incorrect label colors
- Make labelsMaxShown work
- Fix bug when panelNumRows > number of panels
- Fix bug when filter is applied such that some panels are not shown
- Fix toggling of legend items when groups are numeric

### CODE REVIEW

File: theSrc/test/snapshots/travis/master/testPlans/functionalTest/groups-5-snap.png
Comment (JustinCCYap) [inline theSrc/test/snapshots/travis/master/testPlans/functionalTest/groups-5-snap.png:1]: This is a strange example where there are 4 markers but length of the group array is 3. I don't expect this to occur for users in Displayr, so I'm ok to accept this diff.

File: theSrc/scripts/PlotlyChartElements.js [hunk 1]
```diff
@@ -82,7 +82,7 @@ function createPlotlyData (config) {
         }
     } else {
         const indices_by_group = _.groupBy(indices, i => config.group[i])
-        const group_names = Object.keys(indices_by_group)
+        const group_names = _.uniq(config.group)
```
Comment (chschan) [inline theSrc/scripts/PlotlyChartElements.js:85]: What does this change do? Is `Object.keys` removing some values?

File: theSrc/R/htmlwidget.R [hunk 2]
```diff
@@ -444,7 +444,10 @@ CombinedScatter <- function(
     {
         panels <- as.factor(panels)
         panel.labels <- levels(panels)[levels(panels) %in% panels]
-        panels <- as.numeric(panels) - 1
+        panels <- as.numeric(panels)
+        unique_values <- unique(panels)
+        new_values <- rank(unique(panels)) - 1
+        panels <- vapply(panels, function(v) new_values[v == unique_values], numeric(1))
```
Comment (chschan) [inline theSrc/R/htmlwidget.R:457]: Thanks for fixing that. For some reason I thought `as.numeric` of a factor would always start at 1. Maybe just add a comment at the line using `rank` that its to force the values to start at 0

**Review by chschan — APPROVED** [review]

File: theSrc/scripts/PlotlyChartElements.js (hunk 1 above)
Comment (JustinCCYap) [reply theSrc/scripts/PlotlyChartElements.js:85]: It changes the order of `group_names` for an edge case where the groups are numbers "1","1.1", etc. Somehow the previous order was "1","2","3","1.1","1.2",... whereas now it is in the order in which it appears in `config.group`.

**Pull request merged**